airbyte-lib/airbyte_lib/caches/base.py

```diff
@@ -0,0 +1,591 @@
+        return columns
+
+    @final
+    def _get_stream_config(
+        self,
+        stream_name: str,
+    ) -> ConfiguredAirbyteStream:
+        """Return the column definitions for the given stream."""
+        if not self.source_catalog:
+            raise RuntimeError("Cannot get stream JSON schema without a catalog.")
+
+        matching_streams: list[ConfiguredAirbyteStream] = [
+            stream for stream in self.source_catalog.streams if stream.stream.name == stream_name
+        ]
+        if not matching_streams:
+            raise RuntimeError(f"Stream '{stream_name}' not found in catalog.")
+
+        if len(matching_streams) > 1:
+            raise RuntimeError(f"Multiple streams found with name '{stream_name}'.")
+
+        return matching_streams[0]
+
+    @final
+    def _get_stream_json_schema(
+        self,
+        stream_name: str,
+    ) -> dict[str, Any]:
+        """Return the column definitions for the given stream."""
+        return self._get_stream_config(stream_name).stream.json_schema
+
+    @overrides
+    def _write_batch(
+        self,
+        stream_name: str,
+        batch_id: str,
+        record_batch: pa.Table | pa.RecordBatch,
+    ) -> FileWriterBatchHandle:
+        """
+        Process a record batch.
+
+        Return the path to the cache file.
+        """
+        return self.file_writer.write_batch(stream_name, batch_id, record_batch)
+
+    @final
+    @overrides
+    def _finalize_batches(self, stream_name: str) -> dict[str, BatchHandle]:
+        """Finalize all uncommitted batches.
+
+        This is a generic 'final' implementation, which should not be overridden.
+
+        Returns a mapping of batch IDs to batch handles, for those processed batches.
+
+        TODO: Add a dedupe step here to remove duplicates from the temp table.
+              Some sources will send us duplicate records within the same stream,
+              although this is a fairly rare edge case we can ignore in V1.
+        """
+        with self._finalizing_batches(stream_name) as batches_to_finalize:
+            if not batches_to_finalize:
+                return {}
+
+            files: list[Path] = []
+            # Get a list of all files to finalize from all pending batches.
+            for batch_handle in batches_to_finalize.values():
+                batch_handle = cast(FileWriterBatchHandle, batch_handle)
+                files += batch_handle.files
+            # Use the max batch ID as the batch ID for table names.
+            max_batch_id = max(batches_to_finalize.keys())
+
+            # Make sure the target schema and target table exist.
+            self._ensure_schema_exists()
+            final_table_name = self._ensure_final_table_exists(stream_name, create_if_missing=True)
+
+            temp_table_name = self._write_files_to_new_table(files, stream_name, max_batch_id)
+            self._write_temp_table_to_final_table(stream_name, temp_table_name, final_table_name)
```

Comment:
if the schema doesn't match the existing data in the cache, this step will fail, and the `_drop_temp_table` call won't get executed anymore, causing the temporary table to stick around.

Not necessarily a big issue, but I think we should add some logic here to try and clean up any WIP stuff in case something goes wrong.
<img width="270" alt="Screenshot 2024-01-03 at 12 15 48" src="https://github.com/airbytehq/airbyte/assets/1508364/ec2bb940-f8ec-4f32-a3f9-52715bab6346">